### PR TITLE
ARROW-2153/ARROW-2160: [C++/Python]  Fix decimal precision inference

### PR DESCRIPTION
**Pull request by cpcloud**

### CODE REVIEW

File: cpp/src/arrow/python/helpers.cc [hunk 1]
```diff
@@ -99,7 +99,8 @@ Status InferDecimalPrecisionAndScale(PyObject* python_decimal, int32_t* precisio
   DCHECK_NE(precision, NULLPTR);
   DCHECK_NE(scale, NULLPTR);
 
-  OwnedRef as_tuple(PyObject_CallMethod(python_decimal, "as_tuple", "()"));
+  // TODO(phillipc): Make sure we perform PyDecimal_Check(python_decimal) as a DCHECK
```
Comment (cpcloud) [inline cpp/src/arrow/python/helpers.cc]: This requires ARROW-2160 to address.

Comment (wesm) [reply cpp/src/arrow/python/helpers.cc]: Do you want to wait for ARROW-2145 or merge this independently?

Comment (cpcloud) [reply cpp/src/arrow/python/helpers.cc]: I have a couple more tests to add here

**Comment by cpcloud** [comment]
@wesm @pitrou I reintroduced boost regex for parsing the decimal number, to address ARROW-2153. We were using a fairly straightforward algorithm before that implemented the regular expression by hand but it didn't allow numbers like `1e1`. The regular expression to match that would be very complex and hard to read if written by hand so I decided to use boost regex. I didn't go with STL library because it doesn't support named capture groups, which I think make the code much more readable.

**Comment by wesm** [comment]
`std::regex` is totally broken in gcc 4.8.x (what we're using for conda/pip releases) AFAIK so using `<regex>` isn't even an option right now. When we get past gcc 4.8 it might be nice to use the STL regexen

**Comment by cpcloud** [comment]
Yep, I believe `std::regex_match` is implemented as `return false;` :(

**Comment by cpcloud** [comment]
In gcc 4.8 that is. 4.9 may be implemented, but I don't think it supports named capture groups.

**Comment by wesm** [comment]
needs rebase

**Comment by cpcloud** [comment]
Closed in favor of #1651

**Pull request closed without merging**